File: core0/us/sdk/wallet/java/us/wallet/cli/params.java
//===-                           P L E B B L E
//===-                         https://plebble.us
//===-
//===-              Copyright (C) 2017-2022 root1m3@plebble.us
//===-
//===-                      GNU GENERAL PUBLIC LICENSE
//===-                       Version 3, 29 June 2007
//===-
//===-    This program is free software: you can redistribute it and/or modify
//===-    it under the terms of the GPLv3 License as published by the Free
//===-    Software Foundation.
//===-
//===-    This program is distributed in the hope that it will be useful,
//===-    but WITHOUT ANY WARRANTY; without even the implied warranty of
//===-    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
//===-
//===-    You should have received a copy of the General Public License
//===-    along with this program, see LICENCE file.
//===-    see https://www.gnu.org/licenses
//===-
//===----------------------------------------------------------------------------
//===-
package us.wallet.cli;
import us.CFG;                                                                                 // CFG
import static us.gov.crypto.types.*;                                                           // *
import static us.gov.id.types.*;                                                               // *
import static us.gov.io.types.*;                                                               // *
import static us.gov.socket.types.*;                                                           // *
import us.stdint.*;                                                                            // *
import java.io.PrintStream;                                                                    // PrintStream
import us.gov.io.shell_args;                                                                   // shell_args

public class params {

    public static void log(final String line) {     //--strip
        CFG.log_wallet_cli("params: " + line);      //--strip
    }                                               //--strip

    public enum output_mode {
        om_human,
        om_text,
        om_xml,
        om_json,

        num_modes;

        public short as_short() {
            return (short)ordinal();
        }

        public static output_mode from_short(short i) {
            return output_mode.values()[i];
        }
    }

    void constructor() {
        int hc = Runtime.getRuntime().availableProcessors();
        log("hardware concurrency " + hc);  //--strip
        workers = new uint16_t(2 * hc);
        downloads_dir = "/var/www/" + us.CFG.PLATFORM + "_html/downloads";
    }

    public params() {
        log("def constructor"); //--strip
        constructor();
        args = new shell_args("");
    }

    public params(shell_args args_) {
        log("args constructor"); //--strip
        constructor();
        args = args_;
        while (true) {
            String command = args.next_string();
            log("command: " + command); //--strip
            if (command.equals("-wp")) {
                walletd_port.value = args.next_uint16().value;
            }
            else if (command.equals("-lp")) {
                listening_port.value = args.next_uint16().value;
            }
            else if (command.equals("-pp")) {
                published_port.value = args.next_uint16().value;
            }
            else if (command.equals("-pin")) {
                pin.value = args.next_uint16().value;
            }
            else if (command.equals("-whost")) {
                walletd_host.value = args.next_string();
            }
            else if (command.equals("-bp")) {
                backend_port.value = args.next_uint16().value;
            }
            else if (command.equals("-bhost")) {
                backend_host.value = args.next_string();
            }
            else if (command.equals("-home")) {
                homedir = args.next_string();
            }
            else if (command.equals("-d")) {
                daemon = true;
            }
            else if (command.equals("-b")) {
                advanced = false;
            }
            else if (command.equals("-n")) {
                names = true;
            }
            else if (command.equals("-log")) { //--strip
                verbose = true; //--strip
                CFG.logs.set(verbose); //--strip
                log("enabled logs"); //--strip
            } //--strip
            else if (command.equals("-c")) {
                channel.value = args.next_uint16().value;
            }
            else if (command.equals("-v")) {
                max_devices = args.next_uint16();
            }
            else if (command.equals("-e")) {
                max_trade_edges = args.next_uint16();
            }
            else if (command.equals("-w")) {
                workers = args.next_uint16();
            }
            else if (command.equals("-om")) {
                uint16_t m=args.next_uint16();
                if (m.value<output_mode.num_modes.as_short()) om = output_mode.from_short((short)m.value);
            }
            else if (command.equals("-dd")) {
                downloads_dir = args.next_string();
            }
            else if (command.equals("--custodial_wallet_id")) {
                subhome = args.next_string();
            }
            else if (!command.isEmpty()) {
                cmd = command;
                break;
            }
            else {
                break;
            }
        }
    }

    public String get_home_wallet() {
        return get_home() + "/wallet";
    }

    public String get_home_wallet_rpc_client() {
        return get_home_wallet() + "/rpc_client";
    }

    public String get_home() {
        String os = new String();
        if (homedir.isEmpty()) {
            String env = System.getenv("HOME");
            if (env.isEmpty()) {
                System.err.println("No $HOME env var defined");
                System.exit(1);
            }
            os += env + "/" + CFG.DATADIR;
        }
        else {
            os += homedir;
        }
        if (channel.value != 0) {
            os += "/" + channel.value;
        }
        return os;
    }

    public void dump(final String pfx, PrintStream os) {
        os.println(pfx + "channel: " + channel.value);
        os.println(pfx + "home: " + get_home());
        if (CFG.DEBUG == 1) {
            os.println(pfx + "warning: this is a debug build.");
        }
        else {
            os.println(pfx + "this is an optimized build.");
        }

        String strip = "instrumentation code has been stripped.";
        strip = "warning: instrumentation code hasn't been stripped"; //--strip
        os.println(pfx + strip);
        os.println(pfx + "run rpc shell");
        os.println(pfx + "  wallet daemon at: " + walletd_host.value + ":" + walletd_port.value);
        os.println(pfx + "  connect_for_recv  " + rpc__connect_for_recv);
        os.println(pfx + "  stop_on_disconnection  " + rpc__stop_on_disconnection);
    }

    public void connecting_to(PrintStream os) {
        os.print(CFG.PLATFORM);
        os.print("-wallet daemon at " + walletd_host.value + ':' + walletd_port.value);
    }

    public channel_t channel = CFG.CHANNEL;
    public String homedir = new String("");
    public boolean daemon = false;
    public boolean advanced = true;
    public boolean verbose = false;
    public uint16_t max_trade_edges = new uint16_t(4);
    public uint16_t max_devices = new uint16_t(2);
    public uint16_t workers = new uint16_t(4);
    public String subhome = new String("");
    public String cmd = new String("");
    public shell_args args = null;
    public pin_t pin = new pin_t(0);
    public port_t listening_port = CFG.WALLET_PORT;
    public pport_t published_port = CFG.WALLET_PPORT;
    public shost_t walletd_host = new shost_t("localhost");
    public port_t walletd_port = CFG.WALLET_PORT;
    public shost_t backend_host = new shost_t("localhost");
    public port_t backend_port = CFG.PORT;
    public boolean rpc__connect_for_recv = true;
    public boolean rpc__stop_on_disconnection = true;
    public output_mode om = output_mode.om_human;
    public boolean names = false;
    public String downloads_dir = "";
    public boolean banner = true;
}

File: core0/us/wallet/cli/params.cpp
//===-                           P L E B B L E
//===-                         https://plebble.us
//===-
//===-              Copyright (C) 2017-2022 root1m3@plebble.us
//===-
//===-                      GNU GENERAL PUBLIC LICENSE
//===-                       Version 3, 29 June 2007
//===-
//===-    This program is free software: you can redistribute it and/or modify
//===-    it under the terms of the GPLv3 License as published by the Free
//===-    Software Foundation.
//===-
//===-    This program is distributed in the hope that it will be useful,
//===-    but WITHOUT ANY WARRANTY; without even the implied warranty of
//===-    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
//===-
//===-    You should have received a copy of the General Public License
//===-    along with this program, see LICENCE file.
//===-    see https://www.gnu.org/licenses
//===-
//===----------------------------------------------------------------------------
//===-
#include "params.h"

#include <thread>

#include <us/gov/config.h>
#include <us/wallet/engine/wallet_connection_t.h>

//#define loglevel "wallet/cli"
//#define logclass "params"
//#include <us/gov/logs.inc>

// don't log here, it's used before initializing the logging subsystem

using namespace us::wallet::cli;
using c = us::wallet::cli::params;

void c::constructor() {
    workers = 2 * thread::hardware_concurrency();

    ostringstream os;
    os << "/home/" << GOVUSER << "/downloads";
    downloads_dir = os.str();
}

c::params(): args("") {
    constructor();
}

c::params(const shell_args& a): args(a) {
    constructor();
    while(true) {
        string command = args.next<string>();
        if (command == "-wp") {
            walletd_port = args.next<int>();
        }
        else if (command == "-lp") {
            listening_port = args.next<uint16_t>();
        }
        else if (command == "-pp") {
            published_port = args.next<uint16_t>();
        }
        else if (command == "-pin") {
            pin = args.next<uint16_t>();
        }
        else if (command == "-whost") {
            walletd_host = args.next<string>();
        }
        else if (command == "-bp") {
            backend_port = args.next<int>();
        }
        else if (command == "-bhost") {
            backend_host = args.next<string>();
        }
        else if (command == "-home") {
            homedir = args.next<string>();
        }
        else if (command == "-d") {
            daemon = true;
        }
        else if (command == "-b") {
            advanced = false;
        }
        else if (command == "-n") {
            names = false;
        }
        else if (command == "-local") {
            local = true;
        }
//        else if (cmd=="-nc") {
//            p.ncurses=true;
//        }
        else if (command == "-log") {
            verbose = true;
        }
        else if (command == "-nolog") {
            nolog = true;
        }
        else if (command == "-logd") {
            logd = args.next<string>();
        }
        else if (command == "-c") {
            channel = args.next<channel_t>();
        }
        else if (command == "-v") {
            max_devices = args.next<int>();
        }
        else if (command == "-e") {
            max_trade_edges = args.next<int>();
        }
        else if (command == "-w") {
            workers = args.next<int>();
        }
        else if (command == "-om") {
            uint16_t m = args.next<uint16_t>();
            if (m < num_modes) om = (output_mode)m;
        }
        else if (command == "-dd") {
            downloads_dir = args.next<string>();
        }
        else if (command == "-nb") {
            banner = false;
        }
        else if (command == "-B") {
            using wallet_connection_t = us::wallet::engine::wallet_connection_t;
            auto blobb58 = args.next<string>();
            us::wallet::engine::wallet_connection_t wc;
            ko r = wc.read(blobb58);
            if (is_ko(r)) {
                cerr << r << '\n';
            }
            else {
                channel = wc.ip4_endpoint.channel;
                walletd_port = wc.ip4_endpoint.port;
                walletd_host = wc.ip4_endpoint.shost;
            }
        }
        else if (command == "--custodial_wallet_id") {
            subhome = args.next<string>();
        }
        #if CFG_FCGI == 1
        else if (command == "-fcgi") {
           fcgi = true;
        }
        #endif
        else if (command == "-t") {
            rpc_timeout_secs = args.next<uint16_t>(CFG_DGRAM_ROUNDTRIP_TIMEOUT_SECS);
        }
        else if (!command.empty()) {
            cmd = command;
            break;
        }
        else {
            break;
        }
    }
}

string c::get_home_wallet() const {
    return get_home() + "/wallet";
}

string c::get_home_wallet_rpc_client() const {
    return get_home_wallet() + "/rpc_client";
}

string c::get_home_gov() const {
    return get_home() + "/gov";
}

string c::get_home() const {
    ostringstream os;
    if (homedir.empty()) {
        const char* env_p = std::getenv("HOME");
        if (!env_p) {
            cerr << "No $HOME env var defined" << endl;
            exit(1);
        }
        os << env_p << "/" << DATADIR;
    }
    else {
        os << homedir;
    }
    if (channel != 0) {
        os << '/' << channel;
    }
    return os.str();
}

void c::dump(ostream& os) const {
    dump("", os);
}

void c::dump(const string& pfx, ostream& os) const {
    os << pfx << "channel: " << channel << '\n';
    os << pfx << "home: " << get_home() << '\n';
    #ifdef DEBUG
        os << pfx << "warning: this is a debug build.\n";
    #else
        os << pfx << "this is an optimized build.\n";
    #endif
    #if CFG_LOGS == 1
        if (verbose) {
            os << pfx << "writing log to cout\n";
        }
        else {
            os << pfx << "warning: logs are ON. Writing files in " << logd << '\n';
        }
    #else
        os << pfx << "logs: disabled.\n";
    #endif

    if (daemon) {
        os << pfx << "run wallet daemon\n";
        os << pfx << "  gov daemon at: " << backend_host << ":" << backend_port << '\n';
        os << pfx << "  listening port: " << listening_port << '\n';
        os << pfx << "  published listening port: " << published_port << '\n';
        os << pfx << "  max device edges: " << (int)max_devices << '\n';
        os << pfx << "  max trader edges: " << (int)max_trade_edges << '\n';
        os << pfx << "  hardware concurrency: " << thread::hardware_concurrency() << '\n';
        os << pfx << "  workers: " << (int)workers << '\n';
        #if CFG_FCGI == 1
            os << pfx << "  fcgi: " << (int)fcgi << '\n';
            if (fcgi) {
                os << pfx << "       listening at: 127.0.0.1:9000\n";
            }
        #endif
    }
    else {
        os << pfx << "run rpc shell\n";
        os << pfx << "  wallet daemon at: " << walletd_host << ":" << walletd_port << '\n';
        os << pfx << "  connect_for_recv " << rpc__connect_for_recv << '\n';
        os << pfx << "  stop_on_disconnection " << rpc__stop_on_disconnection << '\n';
    }
}

void c::connecting_to(ostream& os) const {
    os << PLATFORM;
    os << "-wallet daemon at " << walletd_host << ':' << walletd_port;
}

